fix(job-server-api,job-server,akka-app): fixed starting a new job does not determine proxy user issue #827

zromi18 opened the pull request:
Issues:
- Starting a new job does not determine proxy user #827

Fixes:
- Added determine proxy user code to adHoc context creation and job starting

Pull Request checklist:
- [X] The commit(s) message(s) follows the contribution guidelines ?
- [X] Tests for the changes have been added (for bug fixes / features) ?
- [No] Docs have been added / updated (for bug fixes / features) ?

Current behavior :
- After context is created by proxy-user then starting a job for such a context fails or the context name has to be prepended with user name and separator (~)

New behavior :
- After context is created by proxy-user then starting a job for such a context with just its name succeeds,
- AdHoc context names are internally prepended with user name and separator (~)

noorul commented:
Can you rebase, squash and push?

zromi18 commented:
@noorul Please review my changes. If you are not happy with my rebase, squash & push efforts, please provide me with more specific quideliness. Thank you.

Meehau commented:
I'm also experiencing this issue. It would be really helpful if we could merge this pull request. @noorul @velvia - can you review this again and, if everything OK, accept it?

velvia commented on README.md:
```diff
@@ -242,6 +244,44 @@ Now let's run the job in the context and get the results back right away:
 
 Note the addition of `context=` and `sync=true`.
 
+### Debug mode
```
Nice.

velvia commented on bin/server_stop.sh:
```diff
@@ -19,14 +19,22 @@ fi
-  "$(dirname "$0")"/kill-process-tree.sh 15 $PID && rm "$pidFilePath"
-  echo '...job server stopped'
+  if ! kill -0 $PID; then
+    echo "PID file exists but the process $PID does not exist. Removing $pidFilePath"
```
Nice

velvia commented on doc/javaapi.md:
````diff
@@ -0,0 +1,41 @@
+###Basic Example
+
+```java
+public class JavaHelloWorldJob implements JSparkJob<String> {
````
Yes, thanks again for adding this, does the main README have a link to this one?

zromi18 replied:
I have added the link to the main README.md

velvia commented on job-server-api/src/main/scala/spark/jobserver/context/SparkContextFactory.scala:
```diff
@@ -130,12 +130,12 @@ trait JavaContextFactory extends SparkContextFactory {
- *
- * If you create your own SparkContextFactory, please make sure it has zero constructor args.
- */
+  * The default factory creates a standard SparkContext.
```
Maybe not a big deal, but I would separate out the style-related changes to make it easier to review....  this could be because I'm just viewing all the changes at once, but that's because I see some commits which are revising earlier ones.

zromi18 replied:
I will keep that in mind for the future.

velvia commented on job-server/src/main/scala/spark/jobserver/WebApi.scala:
```diff
@@ -345,7 +340,9 @@ class WebApi(system: ActorSystem,
           .map { contexts =>
-            ctx.complete(removeProxyUserPrefix(authInfo, contexts)) }
+            ctx.complete(SparkJobUtils.removeProxyUserPrefix(
+              authInfo.toString, contexts,
```
What is `authInfo.toString` if auth is not used, which is the majority use case probably?

zromi18 replied:
If authentication is not used then authInfo is new AuthInfo(new User("anonymous")). I have changed that to be lazy evaluated parameter of SparjJobUtils.removeProxyUserPrefix

velvia commented on job-server/src/main/scala/spark/jobserver/io/FileCacher.scala:
```diff
@@ -29,19 +29,34 @@ trait FileCacher {
+      tempOutFileName, targetFullBinaryName: Any)
+
+    val tempFile = new File(rootDir, tempOutFileName)
+    if( ! tempFile.renameTo(new File(rootDir, targetFullBinaryName))) {
```
I would prefer the style `if (!tempFile.renameTo(....`
This is how most of the rest of the code is.

zromi18 replied:
I have changed the style as per your suggestions.

zromi18 commented:
@velvia thank you for the review. I have addressed most of your suggestions or replied to a few of them.
@noorul could you please make your review as well. As I understand I need acceptance from two reviewers to get my changes merged.

velvia commented on job-server/src/main/scala/spark/jobserver/io/FileCacher.scala:
```diff
@@ -50,10 +50,10 @@ trait FileCacher {
 
     val tempFile = new File(rootDir, tempOutFileName)
-    if( ! tempFile.renameTo(new File(rootDir, targetFullBinaryName))) {
+    if(!tempFile.renameTo(new File(rootDir, targetFullBinaryName))) {
```
Sorry to be nitpicky, but could you please add a space between if and ( ?  Thanks

zromi18 replied:
Done, plus the same formatting fix in WebApi.scala.

noorul replied:
Can you re-base with master as I have pushed some scalastyle config to detect such style issues? Also it will be great if you could squash the commits into one.

noorul commented:
Can you rebase, squash and push?

zromi18 commented:
@noorul done, please review

noorul commented:
@zromi18 Why there are so many commits in your patch? If you rebase you should have only one right?

zromi18 commented:
@noorul I am squashing always to two commits one with my changes and second with merge, but when I push the previous ones are still there. Is there a way to remove them?

noorul commented:
1. sync your master branch with upstream.
2. Do a rebase simple rebase on master
3. Now do a rebase interactive
4. Verify that there is only one commit after the commit on master.
5. Force push

zromi18 commented:
@noorul 2 commits one with changes another with merge. Do you accept or want me to squash more?

noorul commented:
@zromi18 I am not sure why you have a merge commit event after a rebase. Also I think that your master is behind upstream master.

zromi18 commented:
@noorul I have tried different things and ensured that my master is aligned with spark-jobserver/master, but still ending up with two commits. Either please make an exception or reject the whole pull request.

velvia commented:
@zromi18 merged

zromi18 commented:
Thank you all for the support and patience. :-)
